src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,208 @@
+        return quorumApkUpdates[quorumNumber][index];
+    }
+
+    /**
+     * @notice get hash of the apk of `quorumNumber` at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     * @param quorumNumber is the quorum whose ApkHash is being retrieved
+     * @param blockNumber is the number of the block for which the latest ApkHash muust be retrieved
+     * @param index is the provided witness of the onchain index calculated offchain
+     */
+    function getApkHashForQuorumAtBlockNumberFromIndex(uint8 quorumNumber, uint32 blockNumber, uint256 index) external returns (bytes32){
+        ApkUpdate memory quorumApkUpdate = quorumApkUpdates[quorumNumber][index];
+        _validateApkHashForQuorumAtBlockNumber(quorumApkUpdate, blockNumber);
+        return quorumApkUpdate.apkHash;
+    }
+
+	/**
+     * @notice get hash of the apk among all quourums at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     */
+    function getGlobalApkHashAtBlockNumberFromIndex(uint32 blockNumber, uint256 index) external view returns (bytes32){
+        ApkUpdate memory globalApkUpdate = globalApkUpdates[index];
+        _validateApkHashForQuorumAtBlockNumber(globalApkUpdate, blockNumber);
+        return globalApkUpdate.apkHash;
+    }
+    /// @notice Returns the length of ApkUpdates for the provided `quorumNumber`
+    function getQuorumApkHistoryLength(uint8 quorumNumber) external view returns(uint32){
+        return uint32(quorumApkUpdates[quorumNumber].length);
+    }
+    /// @notice Returns the length of ApkUpdates for the global APK
+    function getGlobalApkHistoryLength() external view returns(uint32){
+        return uint32(globalApkUpdates.length);
+    }
+
+    function _processGlobalApkUpdate(BN254.G1Point memory point) internal {
+        // load and store in memory in common case we need to access the length again
+        uint256 globalApkUpdatesLength = globalApkUpdates.length;
+        // update the nextUpdateBlockNumber of the previous update
+        if (globalApkUpdates.length > 0) {
+            globalApkUpdates[globalApkUpdatesLength - 1].nextUpdateBlockNumber = uint32(block.number);
+        }
+
+        // accumulate the given point into the globalApk
+        globalApk = globalApk.plus(point);
+        // add this update to the list of globalApkUpdates
+        ApkUpdate memory latestGlobalApkUpdate;
+        latestGlobalApkUpdate.apkHash = BN254.hashG1Point(globalApk);
+        latestGlobalApkUpdate.updateBlockNumber = uint32(block.number);
+        globalApkUpdates.push(latestGlobalApkUpdate);
+    }
+
+    function _processQuorumApkUpdate(bytes memory quorumNumbers, BN254.G1Point memory point) internal {
+        BN254.G1Point memory apkBeforeUpdate;
+        BN254.G1Point memory apkAfterUpdate;
+
+        for (uint8 i = 0; i < uint8(quorumNumbers.length);) {
```

Comment:
is there a reason the loop incrementor `i` here is a uint8? seems like it increases the risk/potential of an overflow bug without delivering any benefit